File: mycuration/src/main/java/com/phicdy/mycuration/view/fragment/ArticlesListFragment.java
package com.phicdy.mycuration.view.fragment;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.View.OnTouchListener;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AbsListView.OnScrollListener;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemLongClickListener;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import com.melnykov.fab.FloatingActionButton;
import com.phicdy.mycuration.R;
import com.phicdy.mycuration.db.DatabaseAdapter;
import com.phicdy.mycuration.presenter.ArticleListPresenter;
import com.phicdy.mycuration.rss.Article;
import com.phicdy.mycuration.rss.Feed;
import com.phicdy.mycuration.rss.UnreadCountManager;
import com.phicdy.mycuration.tracker.GATrackerHelper;
import com.phicdy.mycuration.util.PreferenceHelper;
import com.phicdy.mycuration.util.TextUtil;
import com.phicdy.mycuration.view.ArticleListView;
import com.phicdy.mycuration.view.activity.InternalWebViewActivity;
import com.phicdy.mycuration.view.activity.TopActivity;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

public class ArticlesListFragment extends Fragment implements ArticleListView {

    private ArticleListPresenter presenter;

    private ListView listView;
    private ArticlesListAdapter articlesListAdapter;
    private static final String OPEN_URL_ID = "openUrl";

    private View footer;
    private FloatingActionButton fab;

    private OnArticlesListFragmentListener listener;
    public interface OnArticlesListFragmentListener {
        boolean onListViewTouchEvent(MotionEvent event);
        void finish();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DatabaseAdapter dbAdapter = DatabaseAdapter.getInstance(getActivity());
        UnreadCountManager unreadManager = UnreadCountManager.getInstance(getActivity());

        // Set feed id and url from main activity
        Intent intent = getActivity().getIntent();
        int feedId = intent.getIntExtra(TopActivity.FEED_ID, Feed.ALL_FEED_ID);
        int curationId = intent.getIntExtra(TopActivity.CURATION_ID,
                ArticleListPresenter.DEFAULT_CURATION_ID);
        intent.putExtra(TopActivity.FEED_ID, feedId);

        // Set swipe direction
        PreferenceHelper prefMgr = PreferenceHelper.getInstance(getActivity());
        prefMgr.setSearchFeedId(feedId);
        int swipeDirectionOption = prefMgr.getSwipeDirection();
        boolean isOpenInternal = prefMgr.isOpenInternal();
        boolean isAllReadBack = prefMgr.getAllReadBack();
        boolean isNewestArticleTop = prefMgr.getSortNewArticleTop();
        presenter = new ArticleListPresenter(feedId, curationId, dbAdapter, unreadManager,
                isOpenInternal, isAllReadBack, isNewestArticleTop, swipeDirectionOption);
        presenter.setView(this);
        presenter.create();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            listener = (OnArticlesListFragmentListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString()
                    + " must implement Article list listener");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_articles_list, container, false);
        listView = (ListView) view.findViewById(R.id.lv_article);
        articlesListAdapter = new ArticlesListAdapter(new ArrayList<Article>());
        listView.setAdapter(articlesListAdapter);
        listView.setEmptyView(view.findViewById(R.id.emptyView));
        footer = inflater.inflate(R.layout.footer_article_list_activity, container, false);
        listView.addFooterView(footer);
        fab = (FloatingActionButton) view.findViewById(R.id.fab);
        setAllListener();
        presenter.createView();
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.resume();
    }

    private void setAllListener() {
        // When an article selected, open this URL in default browser
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view,
                                    int position, long id) {
                Article clickedArticle = articlesListAdapter.getItem(position);
                presenter.onListItemClicked(clickedArticle);
            }
        });

        listView.setOnItemLongClickListener(new OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view,
                                           int position, long id) {
                Article clickedArticle = articlesListAdapter.getItem(position);
                if (clickedArticle != null) {
                    presenter.onListItemLongClicked(clickedArticle);
                }
                return true;
            }
        });

        listView.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                return listener.onListViewTouchEvent(event);
            }
        });

        listView.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                return listener.onListViewTouchEvent(event);
            }
        });

        listView.setOnScrollListener(new OnScrollListener() {
            @Override
            public void onScrollStateChanged(AbsListView view, int scrollState) {
            }

            @Override
            public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
                if (totalItemCount == visibleItemCount) {
                    // All of the items are visible
                    return;
                }
                if (totalItemCount == firstVisibleItem + visibleItemCount) {
                    // Reach end of the list
                    presenter.loadArticles();
                }
            }
        });

        fab.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onFabButtonClicked();
                GATrackerHelper.sendEvent(getString(R.string.scroll_article_list));
            }
        });
    }

    public void onFlying(MotionEvent event1, MotionEvent event2, float velocityX) {
        // Set touched position in articles list from touch event
        int touchedPosition = listView.pointToPosition(
                (int) event1.getX(), (int) event1.getY());
        if (touchedPosition < 0 || touchedPosition > articlesListAdapter.getCount()) return;
        presenter.onFlying(touchedPosition, event1, event2, velocityX);
    }

    public void handleAllRead() {
        presenter.handleAllRead();
    }

    @Override
    public void invalidateView() {
        listView.invalidateViews();
    }

    @Override
    public void showFooter() {
        // It causes the crash of ClassCastException
        // if ListView#addFooterView() is called before ListView#setAdapter()
        if (listView.getAdapter() != null) {
            listView.addFooterView(footer);
        }
    }

    @Override
    public void removeFooter() {
        listView.removeFooterView(footer);
    }

    @Override
    public void addArticle(Article article) {
        articlesListAdapter.add(article);
    }

    @Override
    public void openInternalWebView(@NonNull String url) {
        GATrackerHelper.sendEvent(getString(R.string.tap_article_internal));
        Intent intent = new Intent(getActivity(), InternalWebViewActivity.class);
        intent.putExtra(OPEN_URL_ID, url);
        startActivity(intent);
    }

    @Override
    public void openExternalWebView(@NonNull String url) {
        GATrackerHelper.sendEvent(getString(R.string.tap_article_external));
        Uri uri = Uri.parse(url);
        Intent intent = new Intent(Intent.ACTION_VIEW, uri);
        startActivity(intent);
    }

    @Override
    public void notifyListView() {
        articlesListAdapter.notifyDataSetChanged();
    }

    @Override
    public int size() {
        return articlesListAdapter.getCount();
    }

    @Override
    public void finish() {
        listener.finish();
    }

    @Override
    public Article getItem(int position) {
        return articlesListAdapter.getItem(position);
    }

    @Override
    public int getFirstVisiblePosition() {
        return listView.getFirstVisiblePosition();
    }

    @Override
    public int getLastVisiblePosition() {
        return listView.getLastVisiblePosition();
    }

    @Override
    public void showShareUi(@NonNull String url) {
        if (isAdded()) GATrackerHelper.sendEvent(getString(R.string.share_article));
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_TEXT, url);
        startActivity(intent);
    }

    @Override
    public void scroll(int positionToScroll, int pixelFromTopAfterScroll) {
        if (positionToScroll < 0 || pixelFromTopAfterScroll < 0) return;
        listView.smoothScrollToPositionFromTop(positionToScroll, pixelFromTopAfterScroll);
    }

    @Override
    public boolean isBottomVisible() {
        boolean isLastItemVisible = listView.getLastVisiblePosition() == listView.getAdapter().getCount()-1;
        int chilidCount = listView.getChildCount();
        if (chilidCount < 1) return false;
        View lastItem = listView.getChildAt(chilidCount-1);
        if (lastItem == null) return false;
        boolean isLastItemBottomVisible = lastItem.getBottom() == listView.getHeight();
        return isLastItemVisible && isLastItemBottomVisible;
    }

    /**
     *
     * @author phicdy Display articles list
     */
    private class ArticlesListAdapter extends ArrayAdapter<Article> {

        private ViewHolder holder;

        ArticlesListAdapter(ArrayList<Article> articles) {
            /*
            * @param cotext
            *
            * @param int : Resource ID
            *
            * @param T[] objects : data list
            */
            super(getActivity(), R.layout.articles_list, articles);
        }
        @NonNull
        @Override
        public View getView(int position, View convertView, @NonNull ViewGroup parent) {
            // Use contentView
            View row = convertView;
            if (convertView == null) {
                LayoutInflater inflater = getActivity().getLayoutInflater();
                row = inflater.inflate(R.layout.articles_list, parent, false);
                holder = new ViewHolder();
                holder.articleTitle = (TextView)row.findViewById(R.id.articleTitle);
                holder.articlePostedTime = (TextView) row.findViewById(R.id.articlePostedTime);
                holder.articlePoint = (TextView) row.findViewById(R.id.articlePoint);
                holder.articleUrl = (TextView) row.findViewById(R.id.tv_articleUrl);
                holder.feedTitleView = (TextView) row.findViewById(R.id.feedTitle);
                holder.feedIconView = (ImageView) row.findViewById(R.id.iv_feed_icon);
                row.setTag(holder);
            }else {
                holder = (ViewHolder)row.getTag();
            }
            
            Article article = this.getItem(position);
            if (article != null) {
                holder.articleTitle.setText(article.getTitle());
                holder.articleUrl.setText(article.getUrl());

                // Set article posted date
                SimpleDateFormat format = new SimpleDateFormat(
                        "yyyy/MM/dd HH:mm:ss", Locale.US);
                String dateString = format.format(new Date(article
                        .getPostedDate()));
                holder.articlePostedTime.setText(dateString);

                // Set RSS Feed unread article count
                String hatenaPoint = article.getPoint();
                if(hatenaPoint.equals(Article.DEDAULT_HATENA_POINT)) {
                    holder.articlePoint.setText(getString(R.string.not_get_hatena_point));
                }else {
                    holder.articlePoint.setText(hatenaPoint);
                }

                String feedTitle = article.getFeedTitle();
                if(feedTitle == null) {
                    holder.feedTitleView.setVisibility(View.GONE);
                    holder.feedIconView.setVisibility(View.GONE);
                }else {
                    holder.feedTitleView.setText(feedTitle);
                    holder.feedTitleView.setTextColor(Color.BLACK);

                    String iconPath = article.getFeedIconPath();
                    if (!TextUtil.isEmpty(iconPath) && new File(iconPath).exists()) {
                        Bitmap bmp = BitmapFactory.decodeFile(article.getFeedIconPath());
                        holder.feedIconView.setImageBitmap(bmp);
                    }else {
                        holder.feedIconView.setImageResource(R.drawable.no_icon);
                    }
                }
                holder.articleTitle.setTextColor(Color.BLACK);
                holder.articlePostedTime.setTextColor(Color.BLACK);
                holder.articlePoint.setTextColor(Color.BLACK);

                // Change color if already be read
                if (article.getStatus().equals(Article.TOREAD) || article.getStatus().equals(Article.READ)) {
                    holder.articleTitle.setTextColor(Color.GRAY);
                    holder.articlePostedTime.setTextColor(Color.GRAY);
                    holder.articlePoint.setTextColor(Color.GRAY);
                    holder.feedTitleView.setTextColor(Color.GRAY);
                }
            }
            return row;
        }

        private class ViewHolder {
            TextView articleTitle;
            TextView articlePostedTime;
            TextView articlePoint;
            TextView articleUrl;
            TextView feedTitleView;
            ImageView feedIconView;
        }
    }
}

File: legacy/src/main/java/com/phicdy/mycuration/presentation/view/fragment/ArticlesListFragment.kt
package com.phicdy.mycuration.presentation.view.fragment

import android.app.PendingIntent
import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.ItemTouchHelper.ACTION_STATE_SWIPE
import androidx.recyclerview.widget.ItemTouchHelper.LEFT
import androidx.recyclerview.widget.ItemTouchHelper.RIGHT
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.phicdy.mycuration.articlelist.ArticleListView
import com.phicdy.mycuration.data.preference.PreferenceHelper
import com.phicdy.mycuration.di.GlideApp
import com.phicdy.mycuration.entity.Feed
import com.phicdy.mycuration.legacy.R
import com.phicdy.mycuration.presentation.presenter.ArticleListPresenter
import com.phicdy.mycuration.presentation.view.ArticleRecyclerView
import com.phicdy.mycuration.presentation.view.activity.TopActivity
import com.phicdy.mycuration.tracker.TrackerHelper
import com.phicdy.mycuration.util.bitmapFrom
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.koin.android.scope.currentScope
import org.koin.core.parameter.parametersOf
import java.security.InvalidParameterException
import kotlin.coroutines.CoroutineContext


class ArticlesListFragment : Fragment(), ArticleListView, CoroutineScope {

    companion object {
        const val VIEW_TYPE_ARTICLE = 0
        const val VIEW_TYPE_FOOTER = 1
    }

    private lateinit var job: Job
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    private val presenter: ArticleListPresenter by currentScope.inject {
        val feedId = activity?.intent?.getIntExtra(TopActivity.FEED_ID, Feed.ALL_FEED_ID)
                ?: Feed.ALL_FEED_ID
        val curationId = activity?.intent?.getIntExtra(TopActivity.CURATION_ID,
                ArticleListPresenter.DEFAULT_CURATION_ID)
                ?: ArticleListPresenter.DEFAULT_CURATION_ID
        val query = activity?.intent?.getStringExtra(SearchManager.QUERY) ?: ""
        parametersOf(
                feedId,
                curationId,
                query,
                activity?.intent?.action ?: ""
        )
    }

    private lateinit var recyclerView: ArticleRecyclerView
    private lateinit var articlesListAdapter: SimpleItemRecyclerViewAdapter


    private lateinit var listener: OnArticlesListFragmentListener
    private lateinit var emptyView: TextView

    override val firstVisiblePosition: Int
        get() {
            val manager = recyclerView.layoutManager as LinearLayoutManager
            return manager.findFirstVisibleItemPosition()
        }

    override val lastVisiblePosition: Int
        get() {
            val manager = recyclerView.layoutManager as LinearLayoutManager
            return manager.findLastCompletelyVisibleItemPosition()
        }

    override val isBottomVisible: Boolean
        get() {
            val isLastItemVisible = recyclerView.adapter?.let { lastVisiblePosition == it.itemCount - 1 }
                    ?: false
            val chilidCount = recyclerView.childCount
            if (chilidCount < 1) return false
            val lastItem = recyclerView.getChildAt(chilidCount - 1) ?: return false
            val isLastItemBottomVisible = lastItem.bottom == recyclerView.height
            return isLastItemVisible && isLastItemBottomVisible
        }

    interface OnArticlesListFragmentListener {
        fun finish()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        job = Job()

        // Set swipe direction
        val feedId = activity?.intent?.getIntExtra(TopActivity.FEED_ID, Feed.ALL_FEED_ID)
                ?: Feed.ALL_FEED_ID
        val prefMgr = PreferenceHelper
        prefMgr.setSearchFeedId(feedId)

        presenter.setView(this)
        presenter.create()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            listener = context as OnArticlesListFragmentListener
        } catch (e: ClassCastException) {
            throw ClassCastException(context.toString() + " must implement Article list listener")
        }

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_articles_list, container, false)
        recyclerView = view.findViewById(R.id.rv_article) as ArticleRecyclerView
        emptyView = view.findViewById(R.id.emptyViewArticle) as TextView
        recyclerView.layoutManager = LinearLayoutManager(activity)
        articlesListAdapter = SimpleItemRecyclerViewAdapter()
        recyclerView.adapter = articlesListAdapter
        setAllListener()
        launch {
            presenter.createView()
        }
        return view
    }

    override fun onResume() {
        super.onResume()
        presenter.resume()
    }

    override fun onDestroy() {
        job.cancel()
        super.onDestroy()
    }

    private fun setAllListener() {
        val helper = ItemTouchHelper(object : ItemTouchHelper.Callback() {
            override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
                return ItemTouchHelper.Callback.makeFlag(ACTION_STATE_SWIPE, LEFT or RIGHT)
            }

            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                launch {
                    presenter.onSwiped(direction, viewHolder.adapterPosition)
                }
            }
        })
        helper.attachToRecyclerView(recyclerView)
        recyclerView.addItemDecoration(helper)

        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val manager = recyclerView.layoutManager as LinearLayoutManager
                val lastItemPosition = manager.findLastVisibleItemPosition()
                presenter.onScrolled(lastItemPosition)
            }
        })

    }

    fun onFabButtonClicked() {
        launch {
            presenter.onFabButtonClicked()
        }
    }

    fun handleAllRead() {
        launch {
            presenter.handleAllRead()
        }
    }

    override fun openInternalWebView(url: String, rssTitle: String) {
        TrackerHelper.sendButtonEvent(getString(R.string.tap_article_internal))
        val intent = Intent(Intent.ACTION_SEND)
                .setType("text/plain")
                .putExtra(Intent.EXTRA_TEXT, url)
        val pendingIntent = PendingIntent.getActivity(context, 0, intent, 0)
        activity?.let { activity ->
            val icon = bitmapFrom(activity, R.drawable.ic_share)
            icon?.let {
                val customTabsIntent = CustomTabsIntent.Builder()
                        .setShowTitle(true)
                        .setToolbarColor(ContextCompat.getColor(activity, R.color.background_toolbar))
                        .setActionButton(icon, getString(R.string.share), pendingIntent)
                        .build()
                customTabsIntent.launchUrl(activity, Uri.parse(url))
            }
        }
    }

    override fun openExternalWebView(url: String) {
        TrackerHelper.sendButtonEvent(getString(R.string.tap_article_external))
        val uri = Uri.parse(url)
        val intent = Intent(Intent.ACTION_VIEW, uri)
        startActivity(intent)
    }

    override fun notifyListView() {
        articlesListAdapter.notifyDataSetChanged()
    }

    override fun finish() {
        listener.finish()
    }

    override fun showShareUi(url: String) {
        if (isAdded) TrackerHelper.sendButtonEvent(getString(R.string.share_article))
        val intent = Intent(Intent.ACTION_SEND)
        intent.type = "text/plain"
        intent.putExtra(Intent.EXTRA_TEXT, url)
        startActivity(intent)
    }

    override fun scrollTo(position: Int) {
        recyclerView.smoothScrollToPosition(position)
    }

    override fun showEmptyView() {
        recyclerView.visibility = View.GONE
        emptyView.visibility = View.VISIBLE
        emptyView.text = getText(R.string.no_article)
    }

    override fun showNoSearchResult() {
        recyclerView.visibility = View.GONE
        emptyView.visibility = View.VISIBLE
        emptyView.text = getText(R.string.no_search_result)
    }

    inner class SimpleItemRecyclerViewAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            val holder: RecyclerView.ViewHolder
            holder = when (viewType) {
                VIEW_TYPE_FOOTER -> {
                    val footer = LayoutInflater.from(parent.context)
                            .inflate(R.layout.footer_article_list_activity, parent, false)
                    FooterViewHolder(footer)
                }
                VIEW_TYPE_ARTICLE -> {
                    val view = LayoutInflater.from(parent.context)
                            .inflate(R.layout.articles_list, parent, false)
                    ArticleViewHolder(view)
                }
                else -> throw InvalidParameterException("Invalid view type for article list")
            }
            return holder
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            if (holder is ArticleViewHolder) {
                holder.mView.setOnClickListener {
                    launch {
                        presenter.onListItemClicked(holder.getAdapterPosition())
                    }
                }
                holder.mView.setOnLongClickListener {
                    presenter.onListItemLongClicked(holder.getAdapterPosition())
                    true
                }
                presenter.onBindViewHolder(holder, position)
            }
        }

        override fun getItemViewType(position: Int): Int {
            return presenter.onGetItemViewType(position)
        }

        override fun getItemCount(): Int {
            return presenter.articleSize()
        }

        internal inner class FooterViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
        inner class ArticleViewHolder internal constructor(internal val mView: View) : RecyclerView.ViewHolder(mView) {
            internal val articleTitle: TextView = mView.findViewById(R.id.articleTitle) as TextView
            internal val articlePostedTime: TextView = mView.findViewById(R.id.articlePostedTime) as TextView
            internal val articlePoint: TextView = mView.findViewById(R.id.articlePoint) as TextView
            private val articleUrl: TextView = mView.findViewById(R.id.tv_articleUrl) as TextView
            private val feedTitleView: TextView = mView.findViewById(R.id.feedTitle) as TextView
            private val feedIconView: ImageView = mView.findViewById(R.id.iv_feed_icon) as ImageView

            fun setArticleTitle(title: String) {
                articleTitle.text = title
            }

            fun setArticleUrl(url: String) {
                articleUrl.text = url
            }

            fun setArticlePostedTime(time: String) {
                articlePostedTime.text = time
            }

            fun setNotGetPoint() {
                articlePoint.text = getString(R.string.not_get_hatena_point)
            }

            fun setArticlePoint(point: String) {
                articlePoint.text = point
            }

            fun hideRssInfo() {
                feedTitleView.visibility = View.GONE
                feedIconView.visibility = View.GONE
            }

            fun setRssTitle(title: String) {
                feedTitleView.text = title
                feedTitleView.setTextColor(Color.BLACK)
            }

            fun setRssIcon(path: String) {
                GlideApp.with(this@ArticlesListFragment)
                        .load(path)
                        .placeholder(R.drawable.ic_rss)
                        .circleCrop()
                        .error(R.drawable.ic_rss)
                        .into(feedIconView)
            }

            fun setDefaultRssIcon() {
                feedIconView.setImageResource(R.drawable.ic_rss)
            }

            fun changeColorToRead() {
                val color = ContextCompat.getColor(itemView.context, R.color.text_read)
                articleTitle.setTextColor(color)
                articlePostedTime.setTextColor(color)
                articlePoint.setTextColor(color)
                feedTitleView.setTextColor(color)
            }

            fun changeColorToUnread() {
                val color = ContextCompat.getColor(itemView.context, R.color.text_primary)
                articleTitle.setTextColor(color)
                articlePostedTime.setTextColor(color)
                articlePoint.setTextColor(color)
                feedTitleView.setTextColor(color)
            }
        }
    }
}